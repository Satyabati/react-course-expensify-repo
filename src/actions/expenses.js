import uuid from 'uuid';
import database from '../firebase/firebase';
//ADD_EXPENSE
const addExpense = (expense)=>(
    {
        type:'ADD_EXPENSE',
        expense
    }
)

 const startAddExpenses = (expenseData = {}) =>{
    return (dispatch)=>{
      const {
        description = '',
        notes = '',
        amount = 0,
        createdAt = 0
      } = expenseData;
      const expense = {description,notes, amount,createdAt};
      return database.ref('expenses')
      .push(expense)
      .then((ref) =>{
          dispatch(addExpense({
            id:ref.key,
            ...expense
          }))
        
      })
    }
};

//REMOVE_EXPENSES
const removeExpenses =({id}) =>({
    type:'REMOVE_EXPENSE',
    id
}
)

const startRemoveExpenses = (id='')=>{
    return (dispatch)=>{
      return database.ref(`expenses/${id}`).remove().then(()=>{
               dispatch(removeExpenses({
                   id
               }))
           })
    }
}
//EDIT_EXPENSES
const editExpenses=(id,updates) => (
    {
        type:'EDIT_EXPENSE',
        id,
        updates

    }
)

//SET_EXPENSES
const setExpenses = (expenses) =>({
 type:'SET_EXPENSES',
 expenses

})

const startSetExpenses = () =>{
    
    return (dispatch) =>{
        let expenses = [];
        return database.ref('expenses')
     .once('value').then((snapshot) =>{
      snapshot.forEach((childDataSnapShot) =>{
        expenses.push({
          id:childDataSnapShot.key,
           ...childDataSnapShot.val()
         });
       });
     }).catch((e)=>{
       console.log('Error with data fetching', e)
      }).then(() =>{
          dispatch(setExpenses(expenses))
      });
    }
}

module.exports = {
    addExpense,
    removeExpenses,
    editExpenses,
    startAddExpenses,
    setExpenses,
    startSetExpenses,
    startRemoveExpenses
}