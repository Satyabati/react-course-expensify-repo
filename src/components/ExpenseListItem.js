import React from 'react';
import {NavLink } from "react-router-dom";
import moment from 'moment';
import numeral from 'numeral';
export const  ExpenseListItem = ({dispatch,id,amount,createdAt,description})=>(
    <div>
   <h3>{description}</h3>
   <p>{numeral(amount).format('$0,0.00')} 
        -
      {moment(createdAt).format('MMMM Do,YYYY')}</p>

<NavLink to={`edit/${id}`} className="lin">Edit Expense</NavLink>
</div>
);