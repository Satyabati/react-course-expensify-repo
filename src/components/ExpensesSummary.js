import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary =({expensesTotal,expenseCount}) => {
        let amount =  numeral(expensesTotal).format('$0,0.00');
        let expenseWord = expenseCount>0 && expenseCount ===1 ? 'expense': 'expenses';
        return(
            <div>
           Viewing {expenseCount} {expenseWord} totaling {amount}
            </div>
        )
    
}

const mapStateToProps= (state)=>{
    const visibleExpenses = selectExpenses(state.expenses,state.filters);
    const expensesTotal = getTotalExpenses(visibleExpenses);
    const expenseCount = visibleExpenses.length;
    return {
        expensesTotal,
        expenseCount
    }
}
export default connect(mapStateToProps)(ExpensesSummary);