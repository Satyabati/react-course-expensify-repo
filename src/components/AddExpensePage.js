import React  from "react";
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import {startAddExpenses} from '../actions/expenses';

export class AddExpensePage extends React.Component{
    onSubmit = (expenses) =>{
      this.props.startAddExpenses(expenses);
      this.props.history.push('/');
    };

    render(){
        return (
        <div>
        <h1>Add Expenses</h1>
        <ExpenseForm 
         onSubmit={this.onSubmit}
       />
        </div>
        )
    }
} 
const mapDispatchedToProps = (dispatch) => ({
  startAddExpenses: (expense) => dispatch(startAddExpenses(expense))
});

export default connect(undefined,mapDispatchedToProps )(AddExpensePage) ;