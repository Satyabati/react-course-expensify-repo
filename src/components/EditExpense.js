import React  from "react";
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import {editExpenses} from '../actions/expenses'
import { removeExpenses} from '../actions/expenses';
export class EditExpensePage extends React.Component{
    onSubmit=(expense) =>{
        this.props.editExpenses(this.props.match.params.id,expense);
        this.props.history.push('/');
    }

    onClick = () =>{
        this.props.removeExpenses({
            id:this.props.match.params.id
        });
        this.props.history.push('/');
    }
    render(){
        return(
            <div>
            editing the expense for item with id of {this.props.match.params.id}
             <ExpenseForm 
                expense={this.props.expense}
                onSubmit={this.onSubmit}/>
            <button onClick={this.onClick}>Remove</button>
            </div>
        )
    }
};
const mapStateToProps = (state,props)=>(
    {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    });
const mapDispatchedToProps = (dispatch , props) => ({
    editExpenses: (id,expense) => dispatch(editExpenses(id,expense)),
    removeExpenses: (data) => dispatch(removeExpenses(data))
    });

export default connect(mapStateToProps,mapDispatchedToProps)(EditExpensePage);