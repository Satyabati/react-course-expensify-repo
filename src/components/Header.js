import React  from "react";
import {connect} from 'react-redux';
import { startLogOut} from '../actions/auth'
import { NavLink } from "react-router-dom";
export const Header = ({
    startLogOut
}) =>(
    <header>
    <h1>Expensify</h1>
    <NavLink to="/dashboard" activeClassName="is-active" exact={true}>Dashboard</NavLink>  <br/>
    <NavLink to="/create" activeClassName="is-active" exact={true}>Create Expense Page</NavLink> <br/>
    <button onClick={startLogOut}>Logout</button>
    </header>
)  


const mapDispatchedToProps = (dispatch) => ({
    startLogOut :()=> dispatch(startLogOut())
});

export default connect(undefined,mapDispatchedToProps )(Header);