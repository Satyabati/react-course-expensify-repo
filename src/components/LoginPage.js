import React  from "react";
import {connect} from 'react-redux';
import { startLogin } from  '../actions/auth';
export class LoginPage extends React.Component{
 
    render(){
        return (
        <div>
        <h1>Login Page</h1>
        <button onClick={this.props.startLogin}>Login</button>
        </div>
        )
    }
}

const mapDispatchedToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
    });

export default connect(undefined, mapDispatchedToProps)(LoginPage);