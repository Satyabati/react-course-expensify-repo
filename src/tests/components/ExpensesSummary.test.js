import React from 'react';
import {shallow} from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render 1 expense totaling 89',()=>{
    const wrapper = shallow(<ExpensesSummary 
        expensesTotal={89} 
        expenseCount={1}
            />);
    expect(wrapper).toMatchSnapshot();
})

test('should render 2 expenses totaling 890',()=>{
    const wrapper = shallow(<ExpensesSummary 
        expensesTotal={890} 
        expenseCount={2}
            />);
    expect(wrapper).toMatchSnapshot();
})