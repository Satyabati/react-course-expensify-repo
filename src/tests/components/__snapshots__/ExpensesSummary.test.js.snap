// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1 expense totaling 89 1`] = `
<div>
  Viewing 
  1
   
  expense
   totaling 
  $89.00
</div>
`;

exports[`should render 2 expenses totaling 890 1`] = `
<div>
  Viewing 
  2
   
  expenses
   totaling 
  $890.00
</div>
`;
